sys: Serial Number Arithmetic (RFC1982)

cgundogan opened the pull request:
This PR introduces serial number arithmetic as described in [RFC1982](https://tools.ietf.org/html/rfc1982).

While [RFC1982](https://tools.ietf.org/html/rfc1982) deals with sequence numbers of any size, I concentrate more on the "natural" cases, which include 8-bit, 16-bit, 32-bit and 64-bit. Therefore, and to have a type-safe arithmetic, `seqno[8,16,32,64]_t` and the appropriate `compare` and `addition` operations have been defined.

Do you think this handling of sequence numbers is too over-engineered?

RPL is using this kind of sequence number operations and the old implementation did use macros like in the following:
```
#define NG_RPL_COUNTER_MAX                 (255)
#define NG_RPL_COUNTER_LOWER_REGION        (127)
#define NG_RPL_COUNTER_SEQ_WINDOW          (16)
#define NG_RPL_COUNTER_INIT                (NG_RPL_COUNTER_MAX - NG_RPL_COUNTER_SEQ_WINDOW + 1)

static inline uint8_t NG_RPL_COUNTER_INCREMENT(uint8_t counter)
{
    return ((counter > NG_RPL_COUNTER_LOWER_REGION) ?
            ((counter == NG_RPL_COUNTER_MAX) ? counter = 0 : ++counter) :
            ((counter == NG_RPL_COUNTER_LOWER_REGION) ? counter = 0 : ++counter));
}

static inline bool NG_RPL_COUNTER_IS_INIT(uint8_t counter)
{
    return (counter > NG_RPL_COUNTER_LOWER_REGION);
}

static inline bool NG_RPL_COUNTER_GREATER_THAN_LOCAL(uint8_t A, uint8_t B)
{
    return (((A < B) && (NG_RPL_COUNTER_LOWER_REGION + 1 - B + A < NG_RPL_COUNTER_SEQ_WINDOW))
            || ((A > B) && (A - B < NG_RPL_COUNTER_SEQ_WINDOW)));
}

static inline bool NG_RPL_COUNTER_GREATER_THAN(uint8_t A, uint8_t B)
{
    return ((A > NG_RPL_COUNTER_LOWER_REGION) ? ((B > NG_RPL_COUNTER_LOWER_REGION) ?
                NG_RPL_COUNTER_GREATER_THAN_LOCAL(A, B) : 0) :
            ((B > NG_RPL_COUNTER_LOWER_REGION) ? 1 : NG_RPL_COUNTER_GREATER_THAN_LOCAL(A, B)));
}
```

I believing extracting the logic into a separate header file (like proposed in this PR) and making things type-safe is actually the way to go..

kaspar030 commented on sys/include/seqno.h:
```diff
@@ -0,0 +1,321 @@
+/**
+ * @brief       A 8 bit sequence number
+ */
+typedef struct __attribute__((packed)) {
```
why packed?

authmillenon replied:
Why a struct?

kaspar030 commented on sys/include/seqno.h:
```diff
@@ -0,0 +1,321 @@
+ * @return          1, if s1 > s2
+ * @return          -EINVAL, if comparison of the pair (s1,s2) is undefined
+ */
+static inline int seqno8_compare(seqno8_t *s1, seqno8_t *s2)
```
this function seems rather large. Sure this should be static inline?

kaspar030 commented on sys/include/seqno.h:
```diff
@@ -0,0 +1,321 @@
+ * @return          1, if s1 > s2
+ * @return          -EINVAL, if comparison of the pair (s1,s2) is undefined
+ */
+static inline int seqno16_compare(seqno16_t *s1, seqno16_t *s2)
```
same

kaspar030 commented:
The compare functions are identical apart from the types, correct?

cgundogan commented:
Yes, I couldn't find any better solution to reduce code duplication. Do you have any advice here?

authmillenon commented:
I'm somewhat "confused" (I'm not really, since I probably would do the same, but from a "first impression" standpoint) that the module is called `seqno`, in the documentation the numbers are referred to as "sequence numbers" but the module's and PR's title is "serial numbers"

cgundogan commented:
I actually preffer them to be called sequence numbers instead of serial numbers .. though the RFC talks about serial numbers. Do you have a better name than `seqno` in mind?

authmillenon commented:
(I would prefer a re-titling to "Sequence Number Arithmetics" since to my knowledge the name "Serial Number" is more historic).

kaspar030 commented:
Just seq?

authmillenon commented on sys/include/seqno.h:
```diff
@@ -0,0 +1,321 @@
+#endif
+
+#define SEQNO_SERIAL_BITS(X)    (sizeof(X) * 8)
+#define SEQNO_BOUNDARY(X)       (1ULL << (SEQNO_SERIAL_BITS(X) - 1))
```
Documentation for those defines would be nice, especially boundary is not clear to me (shouldn't this be `((1ULL << SEQNO_SERIAL_BITS(X)) - 1)`)

authmillenon commented:
Ping?

cgundogan commented:
I renamed `seqnoX_t` to `seqX_t` and got rid of the structs. Furthermore, I added the possibility to use a custom space size (any power of two minus 1), instead of limiting to `2^8`, `2^16`, `2^32` and `2^64` only.

authmillenon commented on sys/include/seq.h:
```diff
@@ -0,0 +1,493 @@
+ * @return          1, if s1 > s2
+ * @return          -EINVAL, if comparison of the pair (s1,s2) is undefined
+ */
+static inline int seq8_compares(seq8_t s1, seq8_t s2, uint8_t space)
```
This and the other `_compares` functions are still pretty big for an inline function.

cgundogan replied:
@authmillenon do you propose that I should turn this into an own module, wherein the c code resides?

authmillenon replied:
Yes.

cgundogan commented:
@authmillenon I created a separate seq module and moved the code of `*_adds` and `*_compares` into seq.c

authmillenon commented:
The `*_adds` functions could have stayed `inline` but I'm also fine with them being not `inline`.

cgundogan commented:
I guess the compiler knows best what to inline and what not (:

authmillenon commented:
ACK. Tested for iotlab-m3 and native.

authmillenon commented:
Please squash.

cgundogan commented:
squashed, waiting for travis

authmillenon commented on sys/include/seq.h:
```diff
@@ -0,0 +1,397 @@
+
+/** @} */
+#endif /* SEQ_H */
+
```
https://www.youtube.com/watch?v=e-ORhEE9VVg

cgundogan replied:
:) addressed already

authmillenon commented:
And avr and msp-430 are not happy.

cgundogan commented:
kicked travis again, adjusted the tests and did an amend

authmillenon commented:
The avr and msp430 errors seemed valid to me.

cgundogan commented:
to me also. addresing doxygen warnings right now

cgundogan commented:
finally, travis approves.
